Add stack implementation to the library.

peterd-sky opened the pull request.

peterd-sky committed:
Add stack implementation to the library.

peterd-sky commented on examples/custom-gen/my_program_gen.c:
```diff
@@ -43,6 +43,13 @@ int main(int args, char** argv){
 	cdada_list_push_back(list, &foo_key);
 	fprintf(stdout, "The list has a size of %u, contents:\n", cdada_list_size(list));
 
+	cdada_list_pop_front(list);
```
Debug code. I need to remove this.

peterd-sky commented on src/list.cc:
```diff
@@ -649,7 +649,8 @@ int cdada_list_push_back(cdada_list_t* list, const void* val){
 }
 
-static int cdada_list_pop_(cdada_list_t* list, bool front){
+int cdada_list_pop_(cdada_list_t* list, bool front);
```
Debug code. I need to remove this.

peterd-sky committed:
Remove some debug code and update file headers.

msune commented on examples/Makefile.am:
```diff
@@ -25,3 +25,7 @@ str_LDADD = ../src/.libs/libcdada.a -lstdc++
 bbitmap_LDADD = ../src/.libs/libcdada.a -lstdc++
+
+stack_u32_SOURCES = stack_u32.c
+stack_u32_CPPFLAGS = -I$(top_srcdir)/include/ -D__CDADA_INTERNAL_INCLUDE
```
I just realised `examples/` folder should NOT define `-D__CDADA_INTERNAL_INCLUDE`. Original 2 examples were having it.

Could you add a commit on top to remove them from all of them? Thanks

peterd-sky replied:
Done.

msune commented on src/stack.cc:
```diff
@@ -0,0 +1,395 @@
+	if(unlikely(!val))
+		return CDADA_E_INVALID;
+
+	//NOTE: we don't want std::stack insert "replace semantics", so we return
```
Is this comment applicable to`stack->top()`

peterd-sky replied:
No, I will remove it.

peterd-sky replied:
Done.

msune commented on src/stack.cc:
```diff
@@ -0,0 +1,395 @@
+	if(unlikely(!m || m->magic_num != CDADA_MAGIC))
+		return false;
+
+	//NOTE: we don't want std::stack insert "replace semantics", so we return
```
idem

peterd-sky replied:
No, I will remove it.

peterd-sky replied:
Done.

msune commented on test/benchmark.cc:
```diff
@@ -845,6 +1084,7 @@ int main(int args, char** argv){
 	fprintf(stdout, "Type: uint32_t\n");
 	fprintf(stdout, "============================================================\n\n");
 
+#if 0
```
Dead code?

peterd-sky replied:
No, the `#if 0` code blocks were added to skip all but the stack benchmark tests. I have removed the `#if 0` code.

msune commented on test/benchmark.cc:
```diff
@@ -1008,6 +1299,7 @@ int main(int args, char** argv){
 	fprintf(stdout, "(Note: cdada non-custom containers don't support >256 bytes)\n");
 	fprintf(stdout, "============================================================\n\n");
 
+#if 0
```
Dead code?

peterd-sky replied:
Nope. This was an optimization on my part to run just the stack benchmark. I've removed the following statements.

<pre>#if 0 ... #endif</pre>

peterd-sky replied:
Done.

msune commented:
Peter: kudos

First thank you for this PR. It's really good. There are a couple of minor comments here and there, mostly dead code (I think). Once fixed I will merge.

peterd-sky committed:
This commit has the following changes:
- Add the stack tests to the .travis.yml file
- Changes based on code review comments
- Include the stack header files as part of 'make install'
- Add stack to the cdada generator and build the combined custom code

peterd-sky committed:
Remove the one too many colons added the the cdada-gen line for stack

peterd-sky commented:
All code review comments have been addressed. I also added the stack tests to the .travis.yml file and stack code generation to the cdata-gen utility.

msune commented:
ACK. Thanks for the contribution!
